Plugins Design Refresh

mariusandra opened the pull request:
## Changes

In the spirit of PRs over Issues, here is a proposal to change some design elements on the plugins page:

![2021-01-14 10 58 01](https://user-images.githubusercontent.com/53387/104576029-fb1c2000-5657-11eb-8b1e-6361e4eb6568.gif)

The main reason for this is that very very soon we will need to add plugin ordering (e.g. I want to know that the bigquery export runs after the maxmind ip plugin) and that's really hard to do with a grid of boxes. 

The page is still broken on smaller/mobile screens. I'll fix that later once I get some confirmation that this is a good direction design wise.

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

mariusandra committed:
Convert to long horizontal cards

yakkomajuri commented:
Drive by comment that I like this - also because indeed it's a design that makes ordering easier when that comes

yakkomajuri commented:
Another comment would be that our preview deployments don't have the plugin server - wondering if that'd be useful going forward

yakkomajuri reviewed (commented):
I'm on board with this change - but will let the more design-minded folks have the final say.

Only question would be if we have any preferences for styling in JS vs. in separate files

mariusandra commented:
@yakkomajuri all preview deployments should have the plugin server enabled. I made a PR to address this: https://github.com/PostHog/plugin-server/issues/104

mariusandra committed:
plugin cards in mobile mode

mariusandra commented:
Added support for mobile resolutions. Ready for a better review.

Still to come in another PR - (re)ordering plugins via the interface... and possibly merging the two tabs. 

![2021-01-18 12 21 48](https://user-images.githubusercontent.com/53387/104909300-f0c79200-5987-11eb-80e0-ef6f90e9c292.gif)

paolodamico committed:
simplified loading card

paolodamico committed:
repository rename

paolodamico reviewed (changes requested):
I love the general direction of this, think it's quite useful based on the functionality you're introducing. Here are my comments:
- The name of the plugin is not intuitively clickable (it has no color emphasis, no underline, ...). Further, it can be confusing if you take me to GitHub/NPM when I could've been expecting this to take me to the installation page. What I would do is remove the link from the title and then add an icon next to it to take the user to more information (see sample mockup below, not sure about the copy, it could be just the icon, maybe "Learn more"?; FYI we use the language "View plugin ->" in the plugin drawer, let's be consistent on whatever we decide)
    <img width="700" alt="" src="https://user-images.githubusercontent.com/5864173/104973286-d60e1100-59b9-11eb-8876-d72462534fa9.png">
    - Related to the above, I would make the entire box clickable. If it's not installed you open the installation wizard, else you open the configuration drawer.
- Also from the screenshot above, I would introduce a tag to indicate something is from the PH team, or more importantly not from the PH team. Related to a recent PR from @yakkomajuri.
- Not sure if is under this **PR scope?**, but when I install a plugin, it's not removed from the available list (maybe it's because I clicked cancel after that).
    - Related to that, intuitively I would expect that if I click install and then cancel the drawer instead of saving, the plugin wouldn't be installed. 
- Again unsure on **this PR scope?**, but when you install a plugin but don't set the required configuration, logically the enable switch on the plugins list is disabled. However, I have no indication why. I would add a tooltip explaining what's going on **and how to fix it**.
- I would make the disable plugins button at the top red, to convey it's a destructive operation (maybe even make it more explicit with the trash icon?).

Separating comments particularly related to the custom plugins section,
- As a general comment, I'm thinking it might be best to separate it in a tab. It's not part of the "Plugin Repository" and if we have let's say 100 plugins, I'll never find that stuff.
- I would add a "For Advanced Users" or something like that to the custom plugins section. Not so much to deter less advanced users but rather so that users that don't care about it avoid the distraction.
- I would make more explicit the create your plugin in-app thing. Maybe add the text editor right in there, this will also increase the discoverability of the feature. It's hard to understand the true power of that without going to the docs, and it's confusing that what I need to type is a plugin name.
- Depending on the direction you take above, I can suggest some changes to the way you install custom plugins as the current boxes don't look very friendly.

And a few **non-blocking** improvements:
- This new design lets me see less plugins _above the fold_ (or in a single screen height), seems like a very easy improvement would be to add a simple client-side search bar that fuzzy matches titles and descriptions, we're already loading all plugins to the UI.
- When you click install all buttons are set to loading. This doesn't look great. I would either disable all the other buttons with no loading spinner or even remove them from view altogether.
    <img width="254" alt="" src="https://user-images.githubusercontent.com/5864173/104973508-854ae800-59ba-11eb-9225-8fbd745dcdf7.png">


Finally, context on the changes I did directly:
- If I click on Available and you show me Repository, you're making me think (hat tip @mariusandra 😉), that's why I updated the tab name too. I would've pushed back on this name change to "Repository" as I think repository is not as intuitive to convey _plugins that could be installed_. However, as we also have the quite intuitive "+" button, it seems fine.
- Simplified the loading state for the plugin card, seems more crisp and the `Avatar` component can be animated.
    <img width="800" alt="" src="https://user-images.githubusercontent.com/5864173/104973848-b677e800-59bb-11eb-9aa6-04f580de9321.png">

mariusandra committed:
fix errors

mariusandra committed:
improve design of lists

mariusandra committed:
update "learn more" link in plugin drawer

mariusandra committed:
remove debug code

mariusandra committed:
switch to the "installed" tab after installing

mariusandra commented:
Hey! Thanks for the thorough review! Replying to the points one by one and will split this into multiple comments to address all points after fixing them.

This does expand the scope of the PR from the quick fixes I had in mind to something bigger, but let's roll with it! 🚀 

1. "The name of the plugin is not intuitively clickable" --> I added a "learn more" to the end of the description... and made the drawer look the same.
![image](https://user-images.githubusercontent.com/53387/105025914-3ea6cd80-5a4e-11eb-9473-c283027293b9.png)
![image](https://user-images.githubusercontent.com/53387/105026132-83caff80-5a4e-11eb-8e90-3e944d839c38.png)

1. "I would introduce a tag to indicate something is from the PH team, or more importantly not from the PH team. " -- that's now added thanks to [Yakko's PR](https://github.com/PostHog/posthog/pull/2982). It's only visible on the "repository" tag though, as probably after installation it has little value.
![image](https://user-images.githubusercontent.com/53387/105026399-c391e700-5a4e-11eb-9b71-41f21db2b29e.png)
![image](https://user-images.githubusercontent.com/53387/105026496-e3290f80-5a4e-11eb-96ae-3bfa54b1b8f6.png)

1. "When I install a plugin, it's not removed from the available list (maybe it's because I clicked cancel after that)." -- That's a bug with the names of the plugins. See the last point [here](https://github.com/PostHog/plugin-server/issues/100). So let's tackle it outside this PR.

1. "I would expect that if I click install and then cancel the drawer instead of saving, the plugin wouldn't be installed." -- We open the drawer just to save you a click, but the plugin is already installed. Uninstalling the plugin when clicking "cancel" seems like a lot of power for a "cancel" button. I guess the actual fix here is to switch to the "installed" tab after installation. That is now done. This solves another issues as well. Normally the installed plugin should disappear from the list, opening a drawer for something that's not below the drawer. Moving to the "installed" tab makes this interaction cleaner.

1. "I would make the disable plugins button at the top red" -- I actually wouldn't. Otherwise this will visually take the role of the primary action on the page, which is not. Tertiary destructive actions should be as non-emphasised as possible, at least according to what I read in [Refactoring UI](https://refactoringui.com/book/) 
![image](https://user-images.githubusercontent.com/53387/105027721-626b1300-5a50-11eb-975a-b011a749f1c4.png)
![image](https://user-images.githubusercontent.com/53387/105027733-67c85d80-5a50-11eb-91ce-87bbd6885d71.png)
![image](https://user-images.githubusercontent.com/53387/105027759-7282f280-5a50-11eb-99d2-99dbdfe0549b.png)

(more replies to follow)

mariusandra committed:
move custom plugins to a tab

mariusandra committed:
add advanced features warning

mariusandra committed:
remove button to install

mariusandra committed:
disable install buttons that are not clicked

mariusandra commented:
Second batch:

6. "I would make the entire box clickable" --> I'm not sure actually. There are up to four click targets here and making everything else a click target as well might be somewhat unintuitive. Especially on the install screen if you slightly misclick "learn more" it will already install the plugin, which might not be what you want:
<img width="1164" alt="Screenshot 2021-01-19 at 13 34 13" src="https://user-images.githubusercontent.com/53387/105035371-2d17f280-5a5b-11eb-98e5-3c9b8c766136.png">

7. "When you install a plugin but don't set the required configuration, logically the enable switch on the plugins list is disabled." -- Hmm. You can toggle the switch at all times. There's no `disabled` property on it. You just can't save the form without filling in all the required fields. I'm not sure what to change here. There is a UX issue that it's easy to forget to enable a plugin and perhaps the "save" button should be replaced with "Save & Enable", but that's for another PR.

8. Separating "custom plugins" in a tab. That's done. I'm not sure about the wording though. Feel free to change.
![image](https://user-images.githubusercontent.com/53387/105036705-27bba780-5a5d-11eb-90c2-9c8d26139153.png)

9. I also removed the "+ Install new plugin" button. With the two installation tabs, it's not clear where this should take you (you might expect one and get the other), so best to just remove it. I'm sure people will find the tabs.
![image](https://user-images.githubusercontent.com/53387/105037909-f93ecc00-5a5e-11eb-92b4-f69157300f40.png)

10. "I would add a "For Advanced Users" or something like that to the custom plugins section." --> I put something there, though my copywriting is just deplorable :D.
![image](https://user-images.githubusercontent.com/53387/105037612-86355580-5a5e-11eb-8c81-aa521ad19cad.png)

11.  "I would make more explicit the create your plugin in-app thing. Maybe add the text editor right in there, this will also increase the discoverability of the feature. It's hard to understand the true power of that without going to the docs, and it's confusing that what I need to type is a plugin name. Depending on the direction you take above, I can suggest some changes to the way you install custom plugins as the current boxes don't look very friendly." <-- I would suggest moving this to another PR. Adding the text editor is probably too complicated, as 1) it requires loading in [about 6MB of code](https://app.bundletracker.io/b/nNvYVNSdvztajddgGhaadomuB7okP2gL7OrdIydZ) (see the "monaco editor" part), 2) we need two editors, one for the config, 3) it's not clear what would happen if you save 4) etc... I agree this could be made better, though I'd split this out.

12. "This new design lets me see less plugins above the fold (or in a single screen height), seems like a very easy improvement would be to add a simple client-side search bar that fuzzy matches titles and descriptions, we're already loading all plugins to the UI." --- I think now we're still good, but very soon we will need this. I'd actually like to expand on this scope and make a better filtering system. One which lets you choose between e.g. official vs community plugins, search by tags (data in vs data out), etc. I'd do this in another PR.
 
13. "When you click install all buttons are set to loading". -- All buttons are now disabled. It looks more... uhm... interesting this way. Probably an improvement though.
![2021-01-19 14 11 42](https://user-images.githubusercontent.com/53387/105039161-82a2ce00-5a60-11eb-8948-314ca6a71b8f.gif)

I think that covers all points. I think it's ready for another round of feedback and/or tweaking.

paolodamico committed:
minor tweaks

paolodamico reviewed (approved):
Very cool @mariusandra, thanks for taking the time to address this point-by-point! Looks good to go to me, just made some final micro-tweaks.

Re your points for future improvements: 
1-6, 8, 10, 12. 👌
7. Sorry forgot to clarify, I actually meant the toggle on the plugin card not the drawer, that one is disabled and there's no indication as to why or how to fix it.
9. Ok, though I think intuitively you can have the install button and just take you to the repo tab.
11. Agreed, happy to help on this in a different PR
13. Agreed it does still feel a tad weird, something to improve on later.

mariusandra commented:
7. Ah, I misread then! Hopefully fixed in #3006
